airflow/jobs.py

```diff
@@ -2300,9 +2297,35 @@ def _per_task_process(task, key, ti, session=None):
                 self.log.debug('Adding %s to not_ready', ti)
                 ti_status.not_ready.add(key)
 
-            for task in self.dag.topological_sort():
-                for key, ti in list(ti_status.to_run.items()):
-                    _per_task_process(task, key, ti)
+            non_pool_slots = conf.getint('core', 'non_pooled_backfill_task_slot_count')
+
+            try:
+                for task in self.dag.topological_sort():
+                    for key, ti in list(ti_status.to_run.items()):
+                        if task.task_id != ti.task_id:
+                            continue
+                        if task.pool:
+                            pool = session.query(models.Pool) \
+                                .filter(models.Pool.pool == task.pool) \
+                                .first()
+                            if not pool:
+                                raise PoolNotFound('Unknown pool: {}'.format(task.pool))
+
+                            open_slots = pool.open_slots(session=session)
+                            if open_slots <= 0:
+                                raise NoAvailablePoolSlot(
```

Comment:
so now if there is no enough pool, the backfill will fail or retry?

Reply:
Whenever there is no available slots, backfill would sleep and retry.

Reply:
FYI. This is where retry happens:
https://github.com/apache/airflow/blob/84a7a9e6cb81139a2b62996acbd8579aff576810/airflow/jobs.py#L2484-L2508

Reply:
thanks. lgtm